### ISSUE TITLE
ElasticSearch or OpenSearch

### ISSUE THREAD

**Issue by krabina** [SemanticMediaWiki/SemanticMediaWiki#5540]
@designburo raised this topic about SMW here https://github.com/orgs/Open-CSP/discussions/5

So the general question is: should SMW stay with **ElasticSearch** (which since 7.11 has a changed their license model and is currently at version 8.10) or move to **Amazon OpenSearch** (which is  a fork of Elasticsearch 7.10.2.)

For those who are new to the topic, there are some articles for further reading:
- https://en.wikipedia.org/wiki/Elasticsearch and https://en.wikipedia.org/wiki/OpenSearch_(software)
- https://sematext.com/blog/opensearch-vs-elasticsearch-which-one-is-better-sematext/
- https://medium.com/@bhongaleprathamesh/elasticsearch-vs-opensearch-feature-comparison-and-considerations-cd9c990e1482

My personal opinion is that it should stay with whatever decision the Wikimedia Foundation takes with CirrusSearch, but this seems not to be decided yet:
- https://phabricator.wikimedia.org/T272111 and https://phabricator.wikimedia.org/T280482
- https://wikitech.wikimedia.org/wiki/OpenSearch

**Comment by krabina** [comment]
This will only become valid once we resolve the problem that currently SMW is not compatible with ElasticSearch >7, see https://github.com/SemanticMediaWiki/SemanticMediaWiki/issues/5328

**Comment by Designburo** [comment]
The problem we face with customers is that Elastic Search 7.11 is already end-of-life and they do not accept that.
Elastic Search itself is a very proven solution. For users that do not use Cirrus Search, it would be fine to use ES 8.x. It is in a way all depending on what the Wikimedia Foundation will decide. However we cannot wait that long any more.

**Comment by marijnvanwezel** [comment]
We should also consider supporting both ElasticSearch and OpenSearch. I have already started some work on this (https://github.com/WikibaseSolutions/SemanticMediaWiki/tree/elastic-factory-abstraction).

The fact that still no choice has been made indicates that there are good reasons to choose either (e.g. you dislike Amazon, have infrastructure requirements or have a support pact with ElasticSearch and therefore choose ElasticSearch, or you want to use free software and choose OpenSearch).

**Comment by krabina** [comment]
I totally agree @marijnvanwezel. Thank you for your [PR ](https://github.com/SemanticMediaWiki/SemanticMediaWiki/pull/5549) an thank you @JeroenDeDauw for reviewing it!
It is a great idea to support both.

Btw: a question. I think it is a pity that users have to decide between ElasticStore and an RDF store. Do you think, conceptually it would be possible to have the SMWStore to use ElasticStore (which does not replace the SQLStore) or alternatively use ElasticStore AND an RDFStore? This way SMW could provide a SPARQL endpoint and at the same time have the better search experience with Elastic.

**Comment by marijnvanwezel** [comment]
Hi @krabina, I do not have any experience with how the RDFStore works, but conceptually I think it should be possible. The ElasticStore already uses replication and lives alongside the SQLStore. I don't see why that would not be possible if it were an RDFStore instead.

**Comment by krabina** [comment]
This was just announced on Wikitech-l

> This is an advance notice that the Wikimedia Search Platform team will be migrating the search backend for the CirrusSearch extension [1] to OpenSearch. [2] This is in the early stage of planning, with work tentatively slated to begin in the October 2024 - December 2024 timeframe and continue through mid-2025.
> 
> If you run a MediaWiki server with the CirrusSearch extension, you may want to prepare for a migration in your installation as well. There will be changes to MediaWiki PHP code in order to ensure that CirrusSearch works well with an OpenSearch backend, and there will likely be incompatibilities with the old search backend as extension and connection library code is replaced.
> 
> The Search Platform team is available for questions in case you are investigating the transition and how it may affect your MediaWiki search backend setup. The Search Platform team is available on Libera IRC on the #wikimedia-search channel and it hosts monthly office hours the first Wednesday of the month [3], which are announced on the wikitech-l mailing list.
> 
> Thank you.
> 
> Adam Baso
> 
> [1] https://www.mediawiki.org/wiki/Extension:CirrusSearch
> 
> [2] See https://www.mediawiki.org/wiki/Wikimedia_Search_Platform/Decision_Records/Search_backend_replacement_technology for more information.
> 
> [3] https://wikitech.wikimedia.org/wiki/Search_Platform/Contact#Office_Hours
>

**Comment by jaideraf** [comment]
This is interesting: [Elasticsearch is Open Source, Again](https://www.elastic.co/pt/blog/elasticsearch-is-open-source-again). I wonder what WMF is going to do now 😃

**Comment by jaideraf** [comment]
> I wonder what WMF is going to do now 😃

[WMF has decided to stick with its decision to migrate to OpenSearch](https://phabricator.wikimedia.org/T370661#10220958).

**Comment by MPThLee** [comment]
> MediaWiki 1.44+ is compatible with Opensearch 1.3.

https://www.mediawiki.org/wiki/Extension:CirrusSearch

They finally migrated to OpenSearch
